internal/charger/shelly.go

```diff
@@ -0,0 +1,127 @@
+package charger
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/andig/evcc/api"
+	"github.com/andig/evcc/util"
+	"github.com/andig/evcc/util/request"
+)
+
+// Shelly api homepage
+// https://shelly-api-docs.shelly.cloud/#common-http-api
+
+// shellyRelayResponse provides the evcc shelly charger enabled information
+type shellyRelayResponse struct {
+	Ison bool `json:"ison,omitempty"`
+}
+
+// shellyStatusResponse provides the evcc shelly charger current power information
+type shellyStatusResponse struct {
+	Meters []struct {
+		Power float64 `json:"power,omitempty"`
+	} `json:"meters,omitempty"`
+}
+
+// Shelly charger implementation
+type Shelly struct {
+	*request.Helper
+	uri          string
+	channel      int
+	standbypower float64
+}
+
+func init() {
+	registry.Add("shelly", NewShellyFromConfig)
+}
+
+// NewShellyFromConfig creates a Shelly charger from generic config
+func NewShellyFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		URI          string
+		Channel      int
+		StandbyPower float64
+	}{}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	if cc.URI == "" {
+		return nil, errors.New("missing uri")
+	}
+
+	return NewShelly(cc.URI, cc.Channel, cc.StandbyPower)
+}
+
+// NewShelly creates Shelly charger
+func NewShelly(uri string, channel int, standbypower float64) (*Shelly, error) {
+	c := &Shelly{
+		Helper:       request.NewHelper(util.NewLogger("shelly")),
+		uri:          strings.TrimRight(uri, "/"),
+		channel:      channel,
+		standbypower: standbypower,
+	}
+
+	return c, nil
+}
+
+// Enabled implements the api.Charger interface
+func (c *Shelly) Enabled() (bool, error) {
+	var resp shellyRelayResponse
+	err := c.GetJSON(fmt.Sprintf("%s/relay/%d", c.uri, c.channel), &resp)
+
+	return resp.Ison, err
+}
+
+// Enable implements the api.Charger interface
+func (c *Shelly) Enable(enable bool) error {
+	var resp shellyRelayResponse
+	onoff := map[bool]string{true: "on", false: "off"}
+	err := c.GetJSON(fmt.Sprintf("%s/relay/%d?turn=%s", c.uri, c.channel, onoff[enable]), &resp)
+
+	switch {
+	case err != nil:
+		return err
+	case enable && !resp.Ison:
+		return errors.New("switchOn failed")
+	case !enable && resp.Ison:
+		return errors.New("switchOff failed")
+	default:
+		return nil
+	}
+}
+
+// MaxCurrent implements the api.Charger interface
+func (c *Shelly) MaxCurrent(current int64) error {
+	return nil
+}
+
+// Status implements the api.Charger interface
+func (c *Shelly) Status() (api.ChargeStatus, error) {
+	power, err := c.CurrentPower()
+
+	if power > 0 {
+		return api.StatusC, err
+	}
+	return api.StatusB, err
+}
+
+var _ api.Meter = (*Shelly)(nil)
+
+// CurrentPower implements the api.Meter interface
+func (c *Shelly) CurrentPower() (float64, error) {
+	var resp shellyStatusResponse
+	err := c.GetJSON(fmt.Sprintf("%s/%s", c.uri, "status"), &resp)
+	power := resp.Meters[c.channel].Power
```

Comment:
das gibt ein fröhliches panic wenn der Channel nicht existiert. Bitte nochmal probieren- falls das crasht müsste es noch abgefangen werden.

Reply:
hab's abgefangen und auch den Channel in NewShellyFromConfig mit 0 initialisiert.
Relay channels starten mit 0. Siehe: https://shelly-api-docs.shelly.cloud/#shelly1-1pm-relay-0 
Die relay und meter Werte werden im Status Json als Arrays (relays, meters) zurückgegeben:
```json
{
	"wifi_sta": {
		"connected": true,
		"ssid": "XXXX-WLAN",
		"ip": "192.168.2.99",
		"rssi": -77
	},
	"cloud": {
		"enabled": false,
		"connected": false
	},
	"mqtt": {
		"connected": false
	},
	"time": "16:24",
	"unixtime": 1621088649,
	"serial": 6,
	"has_update": false,
	"mac": "84CCA8B10D73",
	"cfg_changed_cnt": 1,
	"actions_stats": {
		"skipped": 0
	},
	"relays": [{
		"ison": false,
		"has_timer": false,
		"timer_started": 0,
		"timer_duration": 0,
		"timer_remaining": 0,
		"overpower": false,
		"source": "http"
	}],
	"meters": [{
		"power": 0.00,
		"overpower": 0.00,
		"is_valid": true,
		"timestamp": 1621095849,
		"counters": [0.000, 0.000, 0.000],
		"total": 0
	}],
	"inputs": [{
		"input": 0,
		"event": "",
		"event_cnt": 0
	}],
	"temperature": 43.55,
	"overtemperature": false,
	"tmp": {
		"tC": 43.55,
		"tF": 110.39,
		"is_valid": true
	},
	"temperature_status": "Normal",
	"ext_sensors": {},
	"ext_temperature": {},
	"ext_humidity": {},
	"update": {
		"status": "idle",
		"has_update": false,
		"new_version": "20210429-101828/v1.10.4-g3f94cd7",
		"old_version": "20210429-101828/v1.10.4-g3f94cd7"
	},
	"ram_total": 49776,
	"ram_free": 37088,
	"fs_size": 233681,
	"fs_free": 149345,
	"ping_check": true,
	"uptime": 5656
}
```